src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -174,97 +208,187 @@ func (s *fileSystemSource) shardAvailability(
+) {
+	// First bucket the shard time ranges by block size
+	var blockSize time.Duration
+	switch run {
+	case bootstrapDataRunType:
+		blockSize = ns.Options().RetentionOptions().BlockSize()
+	case bootstrapIndexRunType:
+		blockSize = ns.Options().IndexOptions().BlockSize()
+	default:
+		panic(fmt.Errorf("unrecognized run type: %d", run))
+	}
 
-			r, err := readerPool.get()
-			if err != nil {
-				s.log.Errorf("unable to get reader from pool")
-				readersCh <- newShardReadersErr(shard, tr, err)
-				continue
-			}
-			t := xtime.FromNanoseconds(info.BlockStart)
-			openOpts := fs.DataReaderOpenOptions{
-				Identifier: fs.FileSetFileIdentifier{
-					Namespace:  namespace,
-					Shard:      shard,
-					BlockStart: t,
-				},
-			}
-			if err := r.Open(openOpts); err != nil {
-				s.log.WithFields(
-					xlog.NewField("shard", shard),
-					xlog.NewField("time", t.String()),
-					xlog.NewField("error", err.Error()),
-				).Error("unable to open fileset files")
-				readerPool.put(r)
-				readersCh <- newShardReadersErr(shard, tr, err)
-				continue
-			}
-			timeRange := r.Range()
-			if !tr.Overlaps(timeRange) {
-				r.Close()
-				readerPool.put(r)
-				continue
-			}
-			readers = append(readers, r)
+	// Group them by block size
+	groupFn := newShardTimeRangesTimeWindowGroups
+	groupedByBlockSize := groupFn(shardTimeRanges, blockSize)
+
+	// Now enqueue across all shards by block size
+	for _, group := range groupedByBlockSize {
+		readers := make(map[shardID]shardReaders, len(group.ranges))
+		for shard, tr := range group.ranges {
+			shardReaders := s.newShardReaders(ns, readerPool, shard, tr)
+			readers[shardID(shard)] = shardReaders
+		}
+		readersCh <- newTimeWindowReaders(group.ranges, readers)
+	}
+}
+
+func (s *fileSystemSource) newShardReaders(
+	ns namespace.Metadata,
+	readerPool *readerPool,
+	shard uint32,
+	tr xtime.Ranges,
+) shardReaders {
+	readInfoFilesResults := fs.ReadInfoFiles(s.fsopts.FilePathPrefix(),
+		ns.ID(), shard, s.fsopts.InfoReaderBufferSize(), s.fsopts.DecodingOptions())
+	if len(readInfoFilesResults) == 0 {
+		return shardReaders{} // No readers
+	}
+
+	readers := make([]fs.DataFileSetReader, 0, len(readInfoFilesResults))
+	for i := 0; i < len(readInfoFilesResults); i++ {
+		result := readInfoFilesResults[i]
+		if result.Err.Error() != nil {
+			s.log.WithFields(
+				xlog.NewField("shard", shard),
+				xlog.NewField("namespace", ns.ID().String()),
+				xlog.NewField("error", result.Err.Error()),
+				xlog.NewField("timeRange", tr.String()),
+				xlog.NewField("path", result.Err.Filepath()),
+			).Error("fs bootstrapper unable to read info file")
+			continue
+		}
+
+		info := result.Info
+		blockStart := xtime.UnixNano(info.BlockStart).ToTime()
+		if !tr.Overlaps(xtime.Range{
+			Start: blockStart,
+			End:   blockStart.Add(ns.Options().RetentionOptions().BlockSize()),
+		}) {
+			continue
+		}
+
+		r, err := readerPool.get()
+		if err != nil {
+			s.log.Errorf("unable to get reader from pool")
```

Comment:
hey so for these cases of skipping error - does it workout because you don't end up marking them fulfilled and the following bootstrapper (if any) has to take over?

Reply:
Yes exactly, it's a bit of a delicate dance, but that's the only way to defer work to the next bootstrapper.

Reply:
Yeah makes sense. Mind putting a little blurb here indicating the same.